### ISSUE TITLE
UIElement.Measure is very slow.

### ISSUE THREAD

**Issue by deeprobin** [dotnet/wpf#4768]
* .NET Core Version: 5.0.6
* Windows version: 21H1 (Build 19043.928)
* Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes
* Is this bug related specifically to tooling in Visual Studio (e.g. XAML Designer, Code editing, etc...)?: No

 
 
 **Problem description:**
UIElement.Measure this very slow
 
 **Actual behavior:** 
![image](https://user-images.githubusercontent.com/10373452/123769736-e2cc5480-d8c9-11eb-9368-5540156292f4.png) 

 **Expected behavior:**
Fast measuring for WPF rendering.
If there is nothing more that can be optimized in the measure method, it might make sense to cache the measures according to certain parameters, since this is very slow with large visual trees. 
Compared to Windows Forms, Windows Forms is superior in terms of performance here.

Would it not make sense to write the measure method in native C++ somehow to optimize the performance accordingly or is the overhead here too large?
 
 **Minimal repro:**
```xaml
<Grid>
    <Grid.ColumnDefinitions>
        <ColumnDefinition Width="*"/>
        <!-- ... Many column definitions -->
    </Grid.ColumnDefinitions>
    <Grid.RowDefinitions>
        <RowDefinition Heigth="*"/>
        <!-- ... Many row definitions -->
    </Grid.RowDefinitions>
    <Button Grid.Row="0" Grid.Column="0">Test</Button>
    <Button Grid.Row="1" Grid.Column="0">Test</Button>
    <Button Grid.Row="0" Grid.Column="2">Test</Button>
    <Button Grid.Row="2" Grid.Column="3">Test</Button>
    <Button Grid.Row="12" Grid.Column="0">Test</Button>
    <Button Grid.Row="7" Grid.Column="8">Test</Button>
    <!-- ... big visual tree -->
</Grid>```

**Comment by deeprobin** [comment]
![image](https://user-images.githubusercontent.com/10373452/123772721-596a5180-d8cc-11eb-9613-489d9427db4b.png)
This are the hot spots of UIElement.Measure. Would it perhaps be useful to take a closer look at the TextAnalyzer performance-wise?

There is the fastest font renderer in the Rust language. Would make sense to create bindings for it and let [fontdue](https://github.com/mooman219/fontdue) do the work. 
Or what do you think?

**Comment by Symbai** [comment]
Instead of switching from C++ to Rust and still having calls to native code (with allocations and overhead), wouldn't it be better to check if the code can be replaced with C# instead? Now that we have Span<char> etc. I don't know if we need any native code here at all. In the .NET runtime repo they're trying to get rid of all native calls because of performance issues.

**Comment by deeprobin** [comment]
@Symbai You're right about that. Are there any benchmarks on how much performance the native overhead takes away? 
And how does it look with the Ahead-of-time compilation, there should be no overhead in my opinion. Would it make sense to distinguish between AOT and JIT compilation (with e.g. preprocessor expressions)?
What should also be noted, in my opinion, is that performance has already been radically improved in interoperability (See [this article](https://devblogs.microsoft.com/dotnet/improvements-in-native-code-interop-in-net-5-0/)).

**Comment by Symbai** [comment]
WPF does not support AOT. There were benchmarks on .NET repo on each commit where they got rid of native code but I don't have any links. The best would be someone (else, because I have no idea what the code does) re-writes the code in C# and compare both. But then it needs to be reviewed by the new WPF repo owners and they are not very good on that.

**Comment by miloush** [comment]
@deeprobin can you provide a repro sample?

**Comment by IAmTheCShark** [comment]
@miloush I think this should be easily achieveable when you create a large grid view with lots of columns and rows and then scroll thorugh it. When virtualization is enabled, textblocks get lots of updates.


[TextblockPerformance.zip](https://github.com/dotnet/wpf/files/6834621/TextblockPerformance.zip)

**Comment by deeprobin** [comment]
@IAmTheCShark This is exactly what I meant in DataGrid and this makes scrolling very unperformant. And it shouldn't be like that.

[Another sample: WpfApp1.zip](https://github.com/dotnet/wpf/files/6845628/WpfApp1.zip)

**Comment by IAmTheCShark** [comment]
@deeprobin I wrote a small hack that uses reflection to turn on some optimizations.
Just out of curiosity, can you give it a try and see if there are any improvements?

The hack will make Typeface.CheckFastPathNominalGlyphs (https://referencesource.microsoft.com/#PresentationCore/Core/CSharp/System/Windows/Media/Typeface.cs,090d1d0c9c031b63,references) return true alot more often

Paste the following code into your MainWindow constructor after InitializeComponent.
The first line creates a Typeface. If you have different FontFamilies, Styles and so on, you might need to call this multiple times.
`            
           
            Typeface typeFace = new Typeface(FontFamily, FontStyle, FontWeight, FontStretch);

            var tryGetGlyphTypeFaceMethod =
                typeof(Typeface).GetMethod("TryGetGlyphTypeface", BindingFlags.Instance | BindingFlags.NonPublic);

            GlyphTypeface glyphTypeFace = tryGetGlyphTypeFaceMethod.Invoke(typeFace, null) as GlyphTypeface;

            var fontFaceLayoutInfoProperty =
                typeof(GlyphTypeface).GetProperty("FontFaceLayoutInfo", BindingFlags.Instance | BindingFlags.NonPublic);

            var fontFaceLayoutInfo = fontFaceLayoutInfoProperty.GetValue(glyphTypeFace);
            var fontFaceLayoutInfoType = fontFaceLayoutInfo.GetType();
            var typographyAvailabilitiesField = fontFaceLayoutInfoType
                .GetField("_typographyAvailabilities", BindingFlags.NonPublic | BindingFlags.Instance);

            var typographyAvailabilitiesProperty
                = fontFaceLayoutInfoType
                    .GetProperty("TypographyAvailabilities", BindingFlags.Instance | BindingFlags.NonPublic);

            // forces initialization of a few things
            typographyAvailabilitiesProperty.GetValue(fontFaceLayoutInfo);

            typographyAvailabilitiesField.SetValue(fontFaceLayoutInfo, 16);
`

**Comment by deeprobin** [comment]
Hey @IAmTheCShark, I just tried your Reflection hack. It makes a minimal difference for me, but it's not an improvement that drastically improves performance.

Have you started to rewrite the text rendering from C++ to C# or is there already a branch for this.

**Comment by deeprobin** [comment]
> @miloush I think this should be easily achieveable when you create a large grid view with lots of columns and rows and then scroll thorugh it. When virtualization is enabled, textblocks get lots of updates.
> 
> [TextblockPerformance.zip](https://github.com/dotnet/wpf/files/6834621/TextblockPerformance.zip)

I have adapted your solution a bit and added a custom rendering using SkiaSharp and notice that the scrolling is much smoother.

Maybe it would be the best solution in the long run to switch from DWrite to SkiaSharp (first for the text rendering).
[TextblockPerformance.zip - with Skia](https://github.com/dotnet/wpf/files/7627799/TextblockPerformance.zip)
